### PR TITLE
Add attributes to each node

### PR DESCRIPTION
**Pull request by TomasVotruba**
Closes https://github.com/phpstan/phpdoc-parser/issues/11

### CODE REVIEW

File: src/Ast/BaseNode.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\PhpDocParser\Ast;
+
+abstract class BaseNode implements Node
+{
+
+	/** @var array<string, mixed> */
+	protected $attributes = [];
```
Comment (ondrejmirtes) [inline src/Ast/BaseNode.php:9]: Should be private

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
I also expected the parser to fill in those attributes needed for a format-preserving printer :) Right now the nodes will come with zero attributes from the parser.

**Comment by TomasVotruba** [comment]
I'm trying to figure out how CI works here. It seems hidden in some layer that does not allow PHP 8 to run cs :/

File: .editorconfig [hunk 2]
```diff
@@ -0,0 +1,9 @@
+root = true
+
+[*.php]
+charset = utf-8
+end_of_line = lf
+insert_final_newline = true
+trim_trailing_whitespace = true
+indent_style = tab
+indent_size = 4
```
Comment (TomasVotruba) [inline .editorconfig:9]: This avoid tabs/spaces conflicts in files out of the box

image

Comment (ondrejmirtes) [reply .editorconfig:9]: CS is in `/build-cs` and runs only on PHP 7.4 afaik. PHP_CodeSniffer is probably not yet compatible with PHP 8.

File: src/Ast/BaseNode.php [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+abstract class BaseNode implements Node
+{
+
+	/** @var array<string, mixed> */
+	private $attributes = [];
+
+	/**
+	 * @param mixed $value
+	 */
+	public function setAttribute(string $key, $value): void
```
Comment (TomasVotruba) [inline src/Ast/BaseNode.php:14]: I don't get this one. What's wrong here?

```bash
FILE: ...ome/runner/work/phpdoc-parser/phpdoc-parser/src/Ast/BaseNode.php
----------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
----------------------------------------------------------------------
 14 | ERROR | Unknown type hint "string" found for $value
    |       | (Squiz.Commenting.FunctionComment.InvalidTypeHint)
----------------------------------------------------------------------
```

File: .editorconfig (hunk 2 above)
Comment (TomasVotruba) [reply .editorconfig:9]: Yes :/ 
https://github.com/phpstan/phpdoc-parser/pull/57#issuecomment-774729574

File: src/Ast/BaseNode.php [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+	public function hasAttribute(string $key): bool
+	{
+		return array_key_exists($key, $this->attributes);
+	}
+
+	/**
+	 * @param mixed|null $default
+	 * @return mixed|null
+	 */
+	public function getAttribute(string $key, $default = null)
```
Comment (JanTvrdik) [inline src/Ast/BaseNode.php:28]: why do you need the `$default`?

Comment (TomasVotruba) [reply src/Ast/BaseNode.php:28]: Not really, just convention from php-parser. It can be removed

File: .editorconfig (hunk 2 above)
Comment (TomasVotruba) [reply .editorconfig:9]: This package is blocker for PHP 8: https://github.com/consistence/coding-standard/blob/4b4b199b14071c26d6b8064a19a4227c1700a000/composer.json#L14

File: src/Ast/BaseNode.php [hunk 5]
```diff
@@ -0,0 +1,37 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\PhpDocParser\Ast;
+
+abstract class BaseNode implements Node
+{
+
+	/** @var array<string, mixed> */
+	private $attributes = [];
```
Comment (JanTvrdik) [inline src/Ast/BaseNode.php:9]: Why not just `public $attributes = []` without the methods? All other properties on nodes are public as well.

File: .editorconfig (hunk 2 above)
Comment (TomasVotruba) [reply .editorconfig:9]: It seems @vasekPurchart is not maintaining that packages for a while https://github.com/consistence/coding-standard/pull/65

File: src/Ast/BaseNode.php (hunk 5 above)
Comment (ondrejmirtes) [reply src/Ast/BaseNode.php:9]: I like the convenience methods `get`/`set`/`has`, with public property you'd have juggle with ugly `isset` or `array_key_exists` in userland code...

Comment (TomasVotruba) [reply src/Ast/BaseNode.php:9]: See https://github.com/phpstan/phpdoc-parser/pull/57/files/1647b310b0d93d2b070875d96558e721335caa07#diff-1c468e54ea94bf2ee8fdc56af68ed6c184a4a54c3298ab74baad9b6404c98bf1

Also better type declaration support on method types.
Making it `public` would override everything with assigns, which is not desired on attributes (compared to all other properties on nodes)

File: src/Ast/BaseNode.php (hunk 3 above)
Comment (ondrejmirtes) [reply src/Ast/BaseNode.php:14]: Probably a bug in an obsolete PHPCS version. I'm gonna look into this once I'm happy with the code and prepare to merge it 👍

File: .editorconfig (hunk 2 above)
Comment (TomasVotruba) [reply .editorconfig:9]: Resolved in standalone PR https://github.com/phpstan/phpdoc-parser/pull/58

File: src/Ast/BaseNode.php (hunk 3 above)
Comment (TomasVotruba) [reply src/Ast/BaseNode.php:14]: See https://github.com/phpstan/phpdoc-parser/pull/58 first

Comment (ondrejmirtes) [reply src/Ast/BaseNode.php:14]: #58 merged, please rebase this PR so we can see it passes :)

File: src/Ast/BaseNode.php (hunk 4 above)
Comment (ondrejmirtes) [reply src/Ast/BaseNode.php:28]: Yeah, please always return null if the attribute doesn't exist, I don't think we'd want to override it.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Please modify the parser to actually fill in those attributes needed for a format-preserving printer, otherwise it's fine.

**Comment by TomasVotruba** [comment]
Hi Ondra,

life came into way and I have other priorities ATM, so I won't have time to play around printer. I don't have it working in my packages yet either.
But it would be very helpful to allow attributes in the nodes. Rector could drop ~40 classes that it uses to generate attribute-aware nodes and whole code would get clean-up. It would also ease testing for printer.

Would you be ok to merge this PR with attributes only and then later handle the printer? 
If so, I will update the PR with your comments and make CI pass.

**Comment by ondrejmirtes** [comment]
Can you please verify this PR is backwards compatible with your other code? Probably is because you'd just be overriding an existing method, but I wouldn't want to break your existing code by releasing this in 0.4.13.

**Comment by TomasVotruba** [comment]
I've tested it sample and it should be ok: 
https://3v4l.org/D4PkT

If it will break, I'll checkout latest 0.9.x Rector release and relase a new version with compatible code.
I'll be here today so I can patch it promptly.

**Comment by TomasVotruba** [comment]
Rebased on `master` and merged fixup commits

File: src/Ast/BaseNode.php (hunk 4 above)
Comment (TomasVotruba) [reply src/Ast/BaseNode.php:29]: I will :+1:

**Comment by TomasVotruba** [comment]
Ready to review :heavy_check_mark:

**Comment by TomasVotruba** [comment]
I've just ported this branch to Rector and it works well :+1: Ready to merge for me. 

See https://github.com/rectorphp/rector/pull/5841 (PHPStan is not relevant; it's failing, because now the classes are loaded twice to give this PR a priority),

**Comment by ondrejmirtes** [comment]
Yeah, give me a bit of time, I'm doing big refactoring this weekend (https://twitter.com/OndrejMirtes/status/1370778018566832131) related to handling `static` type, I'll look into this once I'm finished.

**Comment by TomasVotruba** [comment]
Sure, good luck with those last test :+1:  They're the worst :D

**Comment by ondrejmirtes** [comment]
Hi, I decided on a bit different approach: https://github.com/phpstan/phpdoc-parser/pull/65

1) Add the methods to the Node interface
2) Use a trait instead of class inheritance.

I'm still crediting you in the commit. Because of the first change, I'll have to release a new major version (0.5). I believe you'll be able to use the PR very easily anyway (https://github.com/rectorphp/rector/pull/5841). Thanks.

**Comment by TomasVotruba** [comment]
Hey, that's good news :+1: 

I'll try to update my PR in Rector

**Pull request closed without merging**